src/dtls_combine.rs

```diff
@@ -0,0 +1,219 @@
+use core::array;
+
+use nom::error::{make_error, ErrorKind};
+use nom::{Err, IResult};
+
+use crate::{
+    parse_dtls_message_handshake, DTLSMessage, DTLSMessageHandshake, DTLSMessageHandshakeBody,
+};
+
+const MAX_FRAGMENTS: usize = 50;
+
+/// Combine the given fragments into one. Returns true if the fragments made a complete output.
+/// The fragments are combined in such a way that the output constitutes a complete DTLSMessage.
+///
+/// Returns `None` if the fragments are not complete.
+///
+/// Errors if:
+///
+/// 1. The output is not big enough to hold the reconstituted messages
+/// 2. Fragments are not of the same type (for example ClientHello mixed with Certificate)
+/// 3. (Total) length field differs between the fragments
+/// 4. Fragment offset/length are not consistent with total length
+/// 5. The DTLSMessageHandshakeBody in the message is not a Fragment
+/// 6. The message_seq differs between the fragments.
+///
+/// Panics if there are more than 50 fragments.
+pub fn combine_dtls_fragments<'a>(
+    fragments: &[DTLSMessageHandshake],
+    out: &'a mut [u8],
+) -> IResult<&'a [u8], Option<DTLSMessage<'a>>> {
+    if fragments.is_empty() {
+        return Ok((&[], None));
+    }
+
+    if fragments.len() > MAX_FRAGMENTS {
+        return Err(Err::Error(make_error(&*out, ErrorKind::TooLarge)));
+    }
+
+    const MESSAGE_HEADER_OFFSET: usize = 12;
+
+    // The header all of the fragments share the same DTLSMessage start, apart from the
+    // fragment information. This goes into the first 12 bytes.
+    if out.len() < MESSAGE_HEADER_OFFSET {
+        // Error case 1
+        return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+    }
+
+    // Helper to iterate the fragments in order.
+    let ordered = Ordered::new(fragments);
+
+    // Investigate each fragment_offset + fragment_length to figure out
+    // the max contiguous range over the fragments.
+    let max = ordered.max_contiguous();
+
+    // Unwrap is OK, because we have at least one item (checked above).
+    let first_handshake = ordered.iter().next().unwrap();
```

Comment:
(minor) The test is not obvious, since it relies on `Ordered` keeping the same number of items as `fragments`. A test here (even if redundant) would not be bad to avoid `unwrap`, and would not cause performance problems.

Reply:
Alright. It's an invariant, in the sense that if I got this assumption wrong, there's something more serious wrong (bug). I generally don't like masking those as errors, as it would appear the user provided incorrect data when it's actually a bug in this code.

But happy to change to error.